File: src/li/klass/fhem/fhem/FHEMWebConnection.java
/*
 * AndFHEM - Open Source Android application to control a FHEM home automation
 * server.
 *
 * Copyright (c) 2011, Matthias Klass or third-party contributors as
 * indicated by the @author tags or express copyright attribution
 * statements applied by the authors.  All third-party contributions are
 * distributed under license by Red Hat Inc.
 *
 * This copyrighted material is made available to anyone wishing to use, modify,
 * copy, or redistribute it subject to the terms and conditions of the GNU GENERAL PUBLIC LICENSE, as published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 * or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU GENERAL PUBLIC LICENSE
 * for more details.
 *
 * You should have received a copy of the GNU GENERAL PUBLIC LICENSE
 * along with this distribution; if not, write to:
 *   Free Software Foundation, Inc.
 *   51 Franklin Street, Fifth Floor
 *   Boston, MA  02110-1301  USA
 */

package li.klass.fhem.fhem;

import android.content.SharedPreferences;
import android.preference.PreferenceManager;
import android.util.Log;
import li.klass.fhem.AndFHEMApplication;
import li.klass.fhem.exception.*;
import org.apache.commons.io.IOUtils;
import org.apache.http.HttpResponse;
import org.apache.http.HttpVersion;
import org.apache.http.auth.AuthScope;
import org.apache.http.auth.UsernamePasswordCredentials;
import org.apache.http.client.methods.HttpGet;
import org.apache.http.conn.ClientConnectionManager;
import org.apache.http.conn.ConnectTimeoutException;
import org.apache.http.conn.scheme.PlainSocketFactory;
import org.apache.http.conn.scheme.Scheme;
import org.apache.http.conn.scheme.SchemeRegistry;
import org.apache.http.conn.ssl.SSLSocketFactory;
import org.apache.http.impl.client.DefaultHttpClient;
import org.apache.http.impl.conn.tsccm.ThreadSafeClientConnManager;
import org.apache.http.params.BasicHttpParams;
import org.apache.http.params.HttpConnectionParams;
import org.apache.http.params.HttpParams;
import org.apache.http.params.HttpProtocolParams;
import org.apache.http.protocol.HTTP;

import java.net.URI;
import java.net.URLEncoder;
import java.security.KeyStore;
import java.text.SimpleDateFormat;
import java.util.Date;

public class FHEMWebConnection implements FHEMConnection {

    public static final int CONNECTION_TIMEOUT = 3000;
    public static final int SOCKET_TIMEOUT = 20000;
    private DefaultHttpClient client;

    public static final String FHEMWEB_URL = "FHEMWEB_URL";
    public static final String FHEMWEB_USERNAME = "FHEMWEB_USERNAME";
    public static final String FHEMWEB_PASSWORD = "FHEMWEB_PASSWORD";

    public static final FHEMWebConnection INSTANCE = new FHEMWebConnection();

    private FHEMWebConnection() {
        HttpParams httpParams = new BasicHttpParams();
        HttpConnectionParams.setConnectionTimeout(httpParams, 3000);
        HttpConnectionParams.setSoTimeout(httpParams, 20000);
        client = createNewHTTPClient();
    }

    @Override
    public String xmllist() {
        return request("xmllist");
    }

    @Override
    public String fileLogData(String logName, Date fromDate, Date toDate, String columnSpec) {
        SimpleDateFormat dateFormat = new SimpleDateFormat("yyyy-MM-dd_HH:mm");
        String command = new StringBuilder().append("get ").append(logName).append(" - - ")
                .append(dateFormat.format(fromDate)).append(" ")
                .append(dateFormat.format(toDate)).append(" ")
                .append(columnSpec).toString();

        return request(command).replaceAll("#" + columnSpec, "").replaceAll("[\\r\\n]", "");
    }

    @Override
    public String executeCommand(String command) {
        return request(command);
    }

    private String request(String command) {


        String content;
        try {
            HttpGet request = new HttpGet();
            String urlString = getURL() + "?XHR=1&cmd=" + URLEncoder.encode(command);
            Log.i(FHEMWebConnection.class.getName(), "accessing URL " + urlString);
            URI uri = new URI(urlString);

            client.getCredentialsProvider().setCredentials(new AuthScope(uri.getHost(), uri.getPort()),
                    new UsernamePasswordCredentials(getUsername(), getPassword()));

            request.setURI(uri);

            HttpResponse response = client.execute(request);
            if (response.getStatusLine().getStatusCode() == 401) {
                throw new AuthenticationException(response.getStatusLine().toString());
            }
            content = IOUtils.toString(response.getEntity().getContent());

            if (content.contains("<title>") || content.contains("<div id=")) {
                throw new FHEMUpdateRequiredException();
            }
        } catch (AndFHEMException e) {
            throw e;
        } catch (ConnectTimeoutException e) {
            throw new TimeoutException(e);
        } catch (Exception e) {
            throw new HostConnectionException(e);
        }

        return content;
    }

    private String getURL() {
        SharedPreferences sharedPreferences = PreferenceManager.getDefaultSharedPreferences(AndFHEMApplication.getContext());
        String url = sharedPreferences.getString(FHEMWEB_URL, null);
        if (url.lastIndexOf("/") == url.length() - 1) {
            return url.substring(0, url.length() -1);
        }
        return url;
    }

    private String getUsername() {
        SharedPreferences sharedPreferences = PreferenceManager.getDefaultSharedPreferences(AndFHEMApplication.getContext());
        return sharedPreferences.getString(FHEMWEB_USERNAME, "");
    }

    private String getPassword() {
        SharedPreferences sharedPreferences = PreferenceManager.getDefaultSharedPreferences(AndFHEMApplication.getContext());
        return sharedPreferences.getString(FHEMWEB_PASSWORD, "");
    }

    private DefaultHttpClient createNewHTTPClient() {
        try {
            KeyStore trustStore = KeyStore.getInstance(KeyStore.getDefaultType());
            trustStore.load(null, null);

            SSLSocketFactory sf = new CustomSSLSocketFactory(trustStore);
            sf.setHostnameVerifier(SSLSocketFactory.ALLOW_ALL_HOSTNAME_VERIFIER);

            HttpParams params = new BasicHttpParams();
            HttpProtocolParams.setVersion(params, HttpVersion.HTTP_1_1);
            HttpProtocolParams.setContentCharset(params, HTTP.UTF_8);
            HttpConnectionParams.setConnectionTimeout(params, CONNECTION_TIMEOUT);
            HttpConnectionParams.setSoTimeout(params, SOCKET_TIMEOUT);

            SchemeRegistry registry = new SchemeRegistry();
            registry.register(new Scheme("http", PlainSocketFactory.getSocketFactory(), 80));
            registry.register(new Scheme("https", sf, 443));

            ClientConnectionManager ccm = new ThreadSafeClientConnManager(params, registry);

            return new DefaultHttpClient(ccm, params);
        } catch (Exception e) {
            return new DefaultHttpClient();
        }
    }
}

File: src/main/java/li/klass/fhem/fhem/FHEMWEBConnection.kt
/*
 * AndFHEM - Open Source Android application to control a FHEM home automation
 * server.
 *
 * Copyright (c) 2011, Matthias Klass or third-party contributors as
 * indicated by the @author tags or express copyright attribution
 * statements applied by the authors.  All third-party contributions are
 * distributed under license by Red Hat Inc.
 *
 * This copyrighted material is made available to anyone wishing to use, modify,
 * copy, or redistribute it subject to the terms and conditions of the GNU GENERAL PUBLIC LICENSE, as published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 * or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU GENERAL PUBLIC LICENSE
 * for more details.
 *
 * You should have received a copy of the GNU GENERAL PUBLIC LICENSE
 * along with this distribution; if not, write to:
 *   Free Software Foundation, Inc.
 *   51 Franklin Street, Fifth Floor
 *   Boston, MA  02110-1301  USA
 */

package li.klass.fhem.fhem

import android.content.Context
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.net.TrafficStats
import com.google.api.client.extensions.android.http.AndroidHttp.newCompatibleTransport
import com.google.api.client.http.GenericUrl
import com.google.api.client.http.HttpHeaders
import com.google.api.client.http.HttpResponseException
import com.google.common.base.Charsets
import com.google.common.base.MoreObjects.firstNonNull
import com.google.common.base.Optional
import com.google.common.collect.ImmutableMap
import com.google.common.io.CharStreams
import de.duenndns.ssl.MemorizingTrustManager
import li.klass.fhem.error.ErrorHolder
import li.klass.fhem.fhem.RequestResultError.*
import li.klass.fhem.fhem.connection.FHEMServerSpec
import li.klass.fhem.util.ApplicationProperties
import li.klass.fhem.util.CloseableUtil.close
import org.slf4j.LoggerFactory
import java.io.*
import java.net.URLEncoder
import java.security.KeyStore
import javax.net.ssl.*

class FHEMWEBConnection(fhemServerSpec: FHEMServerSpec, applicationProperties: ApplicationProperties) : FHEMConnection(fhemServerSpec, applicationProperties) {

    val basicAuthHeaders: HttpHeaders
        get() {
            val password = Optional.fromNullable(serverSpec.password).or("")
            val username = Optional.fromNullable(serverSpec.username).or("")
            return HttpHeaders().setBasicAuthentication(username, password)
        }

    val password: String
        get() = serverSpec.password

    override fun executeCommand(command: String, context: Context): RequestResult<String> {
        LOG.info("executeTask command " + command)

        val urlSuffix = generateUrlSuffix(command)

        var reader: InputStreamReader? = null
        try {
            val response = executeRequest(urlSuffix, context)
            if (response.error != null) {
                return RequestResult(response.error)
            }

            reader = InputStreamReader(response.content, Charsets.UTF_8)
            val content = CharStreams.toString(reader)
            if (content.contains("<title>") || content.contains("<div id=")) {
                LOG.error("found strange content: " + content)
                ErrorHolder.setError("found strange content in URL $urlSuffix: \r\n\r\n$content")
                return RequestResult(INVALID_CONTENT)
            }

            return RequestResult(content)
        } catch (e: Exception) {
            LOG.error("cannot handle result", e)
            return RequestResult(INTERNAL_ERROR)
        } finally {
            close(reader)
        }
    }

    protected fun generateUrlSuffix(command: String): String? {
        var urlSuffix: String? = null
        try {
            urlSuffix = "?XHR=1&cmd=" + URLEncoder.encode(command, "UTF-8")
        } catch (e: UnsupportedEncodingException) {
            LOG.error("unsupported encoding", e)
        }

        return urlSuffix
    }

    fun executeRequest(urlSuffix: String?, context: Context): RequestResult<InputStream> =
            executeRequest(serverSpec.url, urlSuffix, false, context)

    private fun executeRequest(serverUrl: String, urlSuffix: String?, isRetry: Boolean, context: Context): RequestResult<InputStream> {
        var url: String? = null
        TrafficStats.setThreadStatsTag(1)
        try {
            initSslContext(context)

            url = serverUrl + if (urlSuffix?.contains("cmd=") ?: false) {
                val csrfToken = findCsrfToken(serverUrl) ?: ""
                urlSuffix + "&fwcsrf=" + csrfToken
            } else urlSuffix
            LOG.info("accessing URL {}", url)

            val response = newCompatibleTransport()
                    .createRequestFactory().buildGetRequest(GenericUrl(url))
                    .setConnectTimeout(SOCKET_TIMEOUT)
                    .setReadTimeout(SOCKET_TIMEOUT)
                    .setLoggingEnabled(false)
                    .setHeaders(basicAuthHeaders)
                    .execute()

            LOG.debug("response status code is " + response.statusCode)


            return RequestResult(BufferedInputStream(response.content) as InputStream)
        } catch (e: HttpResponseException) {
            val errorResult = handleHttpStatusCode(e.statusCode)
            val msg = "found error " + errorResult!!.error.declaringClass.getSimpleName() + " for " +
                    "status code " + e.statusCode
            LOG.debug(msg)
            ErrorHolder.setError(null, msg)
            return errorResult

        } catch (e: Exception) {
            LOG.info("error while loading data", e)
            return handleError(urlSuffix ?: "", isRetry, url, e, context)
        }

    }

    private fun findCsrfToken(serverUrl: String): String? {
        try {
            val response = newCompatibleTransport()
                    .createRequestFactory().buildGetRequest(GenericUrl(serverUrl + "?room=notExistingJustToLoadCsrfToken"))
                    .setConnectTimeout(SOCKET_TIMEOUT)
                    .setReadTimeout(SOCKET_TIMEOUT)
                    .setLoggingEnabled(false)
                    .setHeaders(basicAuthHeaders)
                    .execute()

            val value = response.headers.getFirstHeaderStringValue("X-FHEM-csrfToken")

            response.content.close()
            return value
        } catch (e: IOException) {
            LOG.info("cannot load csrf token", e)
            return null
        }

    }

    private fun handleError(urlSuffix: String, isRetry: Boolean, url: String?, e: Exception, context: Context): RequestResult<InputStream> {
        setErrorInErrorHolderFor(e, url, urlSuffix)
        return handleRetryIfRequired(isRetry, RequestResult(HOST_CONNECTION_ERROR), urlSuffix, context)
    }

    private fun handleRetryIfRequired(isCurrentRequestRetry: Boolean, previousResult: RequestResult<InputStream>,
                                      urlSuffix: String, context: Context): RequestResult<InputStream> {
        return if (!serverSpec.canRetry() || isCurrentRequestRetry) {
            previousResult
        } else retryRequest(urlSuffix, context)
    }

    private fun retryRequest(urlSuffix: String, context: Context): RequestResult<InputStream> {
        LOG.info("retrying request for alternate URL")
        return executeRequest(serverSpec.alternateUrl, urlSuffix, true, context)
    }

    override fun requestBitmap(relativePath: String, context: Context): RequestResult<Bitmap> {
        val response = executeRequest(relativePath, context)
        if (response.error != null) {
            return RequestResult(response.error)
        }
        try {
            return RequestResult(BitmapFactory.decodeStream(response.content))
        } finally {
            close(response.content)
        }
    }

    private fun initSslContext(context: Context) {
        try {
            val sslContext = SSLContext.getInstance("TLS")
            val memorizingTrustManager = MemorizingTrustManager(context)
            var clientKeys: Array<KeyManager>? = null
            if (serverSpec.clientCertificatePath != null) {
                val clientCertificate = File(serverSpec.clientCertificatePath)
                val clientCertificatePassword = serverSpec.clientCertificatePassword
                if (clientCertificate.exists()) {
                    val keyStore = loadPKCS12KeyStore(clientCertificate, clientCertificatePassword)
                    val keyManagerFactory = KeyManagerFactory.getInstance("X509")
                    keyManagerFactory.init(keyStore, firstNonNull(clientCertificatePassword, "").toCharArray())
                    clientKeys = keyManagerFactory.keyManagers
                }
            }
            sslContext.init(clientKeys, arrayOf<X509TrustManager>(memorizingTrustManager), java.security.SecureRandom())
            HttpsURLConnection.setDefaultSSLSocketFactory(sslContext.socketFactory)
            HttpsURLConnection.setDefaultHostnameVerifier(
                    memorizingTrustManager.wrapHostnameVerifier(HttpsURLConnection.getDefaultHostnameVerifier()))
        } catch (e: Exception) {
            LOG.error("Error initializing HttpsUrlConnection", e)
        }

    }

    @Throws(Exception::class)
    private fun loadPKCS12KeyStore(certificateFile: File, clientCertPassword: String): KeyStore? {
        val keyStore: KeyStore?
        var fileInputStream: FileInputStream? = null
        try {
            keyStore = KeyStore.getInstance("PKCS12")
            fileInputStream = FileInputStream(certificateFile)
            keyStore!!.load(fileInputStream, clientCertPassword.toCharArray())
        } finally {
            close(fileInputStream)
        }
        return keyStore
    }

    companion object {

        val SOCKET_TIMEOUT = 20000
        private val LOG = LoggerFactory.getLogger(FHEMWEBConnection::class.java)
        val STATUS_CODE_MAP: Map<Int, RequestResultError> = ImmutableMap.builder<Int, RequestResultError>()
                .put(400, BAD_REQUEST)
                .put(401, AUTHENTICATION_ERROR)
                .put(403, AUTHENTICATION_ERROR)
                .put(404, NOT_FOUND)
                .put(500, INTERNAL_SERVER_ERROR)
                .build()

        init {
            HttpsURLConnection.setDefaultSSLSocketFactory(NoSSLv3Factory())
        }

        private fun handleHttpStatusCode(statusCode: Int): RequestResult<InputStream>? {

            val error = STATUS_CODE_MAP[statusCode] ?: return null

            LOG.info("handleHttpStatusCode() : encountered http status code {}", statusCode)
            return RequestResult(error)
        }
    }
}
